### PR TITLE
Add support for OSC 10 and 11 to set the default colors

### LINKED ISSUE #317: Please finish support for OSC 4 and friends
I was really happy to see support for OSC 4 to **set** the 16 palette colors (and I recently helped a friend add it to the docs, _the way it is now_) ... 

However, I have some problems. I'm hoping you can address them all as one request, but I'm happy to split them out. First, as background the [xterm control sequences docs](http://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h2-Operating-System-Commands). Obviously these are not sacred, but they _do_ represent a sort-of standard in that _other_ terminals and tmux have implemented these same sequences the same way...

1. OSC 4 should support multiple colors. XTerm and tmux (and others) support multiple _pairs_ of `index;spec` without having to repeat the `ESC]4;` part, so you can do: `ESC]4;0;rgb:00/00/00;1;rgb:00/00/70;...\\` and list out the whole set of colors as one giant escape sequence.
2. It doesn't allow setting the extended xterm 88 or 256 color palette (see #313).
3. It doesn't support querying. I should be able to put a "?" in place of the rgb spec and get back a control sequence (which is supposed to be the exact same format), such that it could be used to set the color to the same RGB value it is already. This would allow me to _check_ if I'm dealing with PowerShell blue instead of magenta by writing `ESC]4;5;?\\` and parsing the output

Additionally ... without support for OSC 10 and 11 I cannot change the default foreground and background colors, so there's no way to _fix_ (even temporarily) the fact that PowerShell is using 5 as it's default background color. NOTE that these should _also_ support querying with a "?"

## AND EVEN WORSE ...

We really also need support for OSC 104 (and 110 and 111) which would **reset** the colors changed by OSC 4, 10, and 11 to their defaults. Note that if done right, they support resetting individual colors, but as a first draft, I'd settle for the parameterless reset of the entire table -- and for my own purposes, if you implement 110 and 111 as meaning force WHITE text on a BLACK background, _without_ implementing 10 and 11, I wouldn't cry 😉

### PR DESCRIPTION
**Pull request by Jaykul**
## Add support for OSC 10 and 11 to set the default colors

Supports setting the rgb value for the default foreground and background (but no support for querying yet, same as OSC 4)

## PR Checklist
* [x] Closes #317
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [x] Requires documentation to be updated
* [x] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #317

This adds support for OSC 10 to set the default foreground color and OSC 11 to set the default background color (just as if they were set in the profiles.json, except that the settings are not changes, so they can be reloaded).

I have not implemented querying (which is also mentioned in #317), so this only addresses one item in the list.

I have not really added any unit tests -- I'm happy to do so, but unsure where they should go, or what they look like. I found some mocks for the other OSC codes, and implemented those, but didn't spot tests.

### CODE REVIEW

File: src/host/getset.cpp [hunk 1]
```diff
@@ -2170,3 +2170,61 @@ HRESULT DoSrvPrivateSetColorTableEntry(const short index, const COLORREF value)
...
+        if (g.pRender && !gci.IsInVtIoMode())
+        {
+            g.pRender->TriggerRedrawAll();
+        }
+
+        return S_OK;
+    }
+    CATCH_RETURN();
+
+}
```
Comment (flug) [inline src/host/getset.cpp]: cariage return

File: src/host/outputStream.cpp [hunk 2]
```diff
@@ -761,3 +761,14 @@ BOOL ConhostInternalGetSet::PrivateSetColorTableEntry(const short index, const C
...
+
+BOOL ConhostInternalGetSet::PrivateSetDefaultForeground(const COLORREF value) const noexcept
+{
+    return SUCCEEDED(DoSrvPrivateSetDefaultForegroundColor(value));
+}
+
+BOOL ConhostInternalGetSet::PrivateSetDefaultBackground(const COLORREF value) const noexcept
+{
+    return SUCCEEDED(DoSrvPrivateSetDefaultBackgroundColor(value));
+}
```
Comment (flug) [inline src/host/outputStream.cpp]: cariage return

**Comment by vapier** [comment]
i don't think this correctly handles chaining to following OSC sequences.  if you have `\x1b]10;0;0\a`, that should be the same as if you had `\x1b]10;0\a` and `\x1b]11;0\a`.  seeing as how OSC 10/11 are from xterm, and xterm defines that chaining behavior, seems like it should be supported.  it would chain further to OSC 12 and such if they were implemented ...

at a glance, looks like OSC 4 isn't quite handled well either, but i think that's a known bug.

File: src/host/getset.cpp [hunk 3]
```diff
@@ -2170,3 +2170,60 @@ HRESULT DoSrvPrivateSetColorTableEntry(const short index, const COLORREF value)
...
+        // No need to force a redraw in pty mode.
+        if (g.pRender && !gci.IsInVtIoMode())
+        {
+            g.pRender->TriggerRedrawAll();
+        }
+
+        return S_OK;
+    }
+    CATCH_RETURN();
+
```
Comment (HBelusca) [inline src/host/getset.cpp]: Extra newline.

File: src/terminal/adapter/ITermDispatch.hpp [hunk 4]
```diff
@@ -65,6 +65,9 @@ class Microsoft::Console::VirtualTerminal::ITermDispatch
     virtual bool EnableAnyEventMouseMode(const bool fEnabled) = 0; // ?1003
     virtual bool EnableAlternateScroll(const bool fEnabled) = 0; // ?1007
     virtual bool SetColorTableEntry(const size_t tableIndex, const DWORD dwColor) = 0; // OSCColorTable
+    virtual bool SetDefaultForeground(const DWORD dwColor) = 0; // OSCDefaultForeground
+    virtual bool SetDefaultBackground(const DWORD dwColor) = 0; // OSCDefaultBackground
+
```
Comment (HBelusca) [inline src/terminal/adapter/ITermDispatch.hpp]: Extra newline.

**Comment by DHowett-MSFT** [comment]
@Jaykul hey, congrats on being the first external person to spelunk through terminal, conpty _and_ conhost to add a new feature! Thanks, this is really cool.

I'm requesting review from the two people who know the area best. :smile:

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 5]
```diff
@@ -214,6 +185,43 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
         _desiredFont = { _actualFont };
     }
 
+    void TermControl::_BackgroundColorChanged(const uint32_t color)
```
Comment (zadjii-msft) [inline src/cascadia/TerminalControl/TermControl.cpp]: nit: needs a doc comment

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 6]
```diff
@@ -214,6 +185,43 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
...
+                _settings.DefaultBackground(ARGB(0, R, G, B));
+            }
+            else
+            {
+                Media::SolidColorBrush solidColor{};
+                solidColor.Color(bgColor);
+                _root.Background(solidColor);
+                _settings.DefaultBackground(RGB(R, G, B));
+            }
+            });
```
Comment (zadjii-msft) [inline src/cascadia/TerminalControl/TermControl.cpp]: nit: indented too far

File: src/cascadia/TerminalCore/TerminalApi.cpp [hunk 7]
```diff
@@ -169,3 +169,22 @@ bool Terminal::SetColorTableEntry(const size_t tableIndex, const DWORD dwColor)
     _buffer->GetRenderTarget().TriggerRedrawAll();
     return true;
 }
+
+bool Terminal::SetDefaultForeground(const COLORREF dwColor)
```
Comment (zadjii-msft) [inline src/cascadia/TerminalCore/TerminalApi.cpp]: These two methods need doc comments

File: src/terminal/adapter/adaptDispatch.cpp [hunk 8]
```diff
@@ -1853,6 +1853,45 @@ bool AdaptDispatch::SetColorTableEntry(const size_t tableIndex,
     return fSuccess;
 }
 
+bool Microsoft::Console::VirtualTerminal::AdaptDispatch::SetDefaultForeground(const DWORD dwColor)
```
Comment (zadjii-msft) [inline src/terminal/adapter/adaptDispatch.cpp]: nit: more doc comments here

File: src/terminal/parser/OutputStateMachineEngine.cpp [hunk 9]
```diff
@@ -724,6 +726,16 @@ bool OutputStateMachineEngine::ActionOscDispatch(const wchar_t /*wch*/,
             fSuccess = _dispatch->SetColorTableEntry(tableIndex, dwColor);
             TermTelemetry::Instance().Log(TermTelemetry::Codes::OSCCT);
             break;
+
+        case OscActionCodes::SetForegroundColor:
+            fSuccess = _dispatch->SetDefaultForeground(dwColor);
+            TermTelemetry::Instance().Log(TermTelemetry::Codes::OSCCT);
```
Comment (zadjii-msft) [inline src/terminal/parser/OutputStateMachineEngine.cpp]: These two should probably get their own entries in `TermTelemetry`, instead of overloading `TermTelemetry::Codes::OSCCT`

File: src/terminal/parser/OutputStateMachineEngine.hpp [hunk 10]
```diff
@@ -131,8 +131,13 @@ namespace Microsoft::Console::VirtualTerminal
             SetIconAndWindowTitle = 0,
             SetWindowIcon = 1,
             SetWindowTitle = 2,
+            SetWindowProperty = 3, // Not implemented
             SetColor = 4,
+            SetForegroundColor = 10, // WIP
```
Comment (zadjii-msft) [inline src/terminal/parser/OutputStateMachineEngine.hpp]: These two WIP comments seem out of date now :P

Comment (Jaykul) [reply src/terminal/parser/OutputStateMachineEngine.hpp]: Oops. You're right. I totally forgot to remove those. They were there for the sake of a livestreaming session 😊

File: src/terminal/parser/OutputStateMachineEngine.cpp (hunk 9 above)
Comment (Jaykul) [reply src/terminal/parser/OutputStateMachineEngine.cpp]: Added OSCFG and OSCBG to the enum and added them to the WriteFinalTraceLog -- is that all there is?

Comment (zadjii-msft) [reply src/terminal/parser/OutputStateMachineEngine.cpp]: Yep!

File: src/cascadia/TerminalCore/Terminal.cpp [hunk 11]
```diff
@@ -442,6 +442,12 @@ void Terminal::SetScrollPositionChangedCallback(std::function<void(const int, co
     _pfnScrollPositionChanged = pfn;
 }
 
+void Terminal::SetBackgroundCallback(std::function<void(const uint32_t)> pfn) noexcept
+{
+    _pfnBackgroundColorChanged = pfn;
+}
+
+
```
Comment (flug) [inline src/cascadia/TerminalCore/Terminal.cpp]: extra line

File: src/host/ut_host/ScreenBufferTests.cpp [hunk 12]
```diff
@@ -2502,6 +2506,60 @@ void ScreenBufferTests::SetColorTableThreeDigits()
...
+
+    SCREEN_INFORMATION& mainBuffer = gci.GetActiveOutputBuffer();
+    VERIFY_IS_FALSE(mainBuffer._IsAltBuffer());
+    WI_SetFlag(mainBuffer.OutputMode, ENABLE_VIRTUAL_TERMINAL_PROCESSING);
+    VERIFY_IS_TRUE(WI_IsFlagSet(mainBuffer.OutputMode, ENABLE_VIRTUAL_TERMINAL_PROCESSING));
+
+    StateMachine& stateMachine = mainBuffer.GetStateMachine();
+
+    const COLORREF originalColor = gci.GetDefaultBackgroundColor();
+    const COLORREF testColor = RGB(0x33, 0x66, 0x99);
```
Comment (DHowett-MSFT) [inline src/host/ut_host/ScreenBufferTests.cpp]: just on a lark: can we test things that go outside the 0-255 or 0-ff (?) range? some negative cases and some cases for longer inputs might be interesting to see

Comment (Jaykul) [reply src/host/ut_host/ScreenBufferTests.cpp]: Sure, but the parser for _that_ isn't my code (I re-used the existing color parser).

**Comment by Jaykul** [comment]
@vapier wrote:
> i don't think this correctly handles chaining to following OSC sequences.

That's correct. It doesn't handle chaining at all. It also doesn't handle color names.

Since this was my first PR, I was trying not to rewrite, but to just add a little on top of what was there and learn how things work -- I even reused the color parser that was written for OSC 4, though it won't support color names, meaning you can only set RGB values, and not just reuse colors from your base 16 for foreground and background.

The structure of the OSC handler in the State Machine seems designed on the assumption that they are single-valued (meaning one value per escape sequence), which makes chaining basically impossible.

Basically, still to do:
1. Rewrite the `ActionOscDispatch` to support chains of values
2. Make OSC 4 support setting the whole color palette at once
3. Make OSC 10, 11, 12 support setting all of them at once
4. Add support for lookups ("?") to output the current value
5. Add support for resetting colors (OSC 104..119)
6. Rewrite the color parser to include x11 color names

Those things are definitely needed, but didn't seem necessary for a first pass.

Somewhere along the way someone will need to decide if we're going to do _special colors_ for  bold/italic/underline/blink/reverse ... and which window properties should be settable by OSC 3 😉 @oising?

**Comment by Jaykul** [comment]
This latest build is failing `TestDbcsBisectWriteCellsBeginW` but I'm not sure what that is.
I'll have to look at it tomorrow.

File: src/host/ut_host/ScreenBufferTests.cpp (hunk 12 above)
Comment (Jaykul) [reply src/host/ut_host/ScreenBufferTests.cpp]: Added just a couple. How deep do you want to go? They're all just going to silently do nothing unless we find a bug in the parsers.

There should probably be (maybe there are) unit tests for the color parser already?

**Comment by vapier** [comment]
I'm not sure if missing support for multiple parameters is worse than not supporting it at all. Not saying you need to fix the world before making any progress, but certainly both multi value sequences (OSC 4 & OSC 10+) need fixing at some point before the VT parser could be considered reliable.

I'm just making observations from the sidelines ... I'm not a developer for this project ;).

**Comment by zadjii-msft** [comment]
> I'm not sure if missing support for multiple parameters is worse than not supporting it at all. 

I will say as a developer for this project and say that missing support for multiple parameters is **not** worse than not supporting it at all. 

While we don't technically have a _complete_ parser without support for multi-param OSC sequences, we definitely have a good enough parser without them. While we won't reliably work with 100% of scenarios, we do work well enough with 99% of scenarios that are actually used. We've been working on the VT implementation iteratively, and getting better with each release. Someday soon we may even have 100% compatibility. 

I've moved this discussion to #942, so we can track adding support there.

<hr>
@jaykul That's one of those tests that's just a little flaky. I'll re-run the CI, it should go away.

File: src/cascadia/TerminalCore/TerminalDispatch.cpp [hunk 13]
```diff
@@ -70,3 +70,14 @@ bool TerminalDispatch::SetColorTableEntry(const size_t tableIndex,
 {
     return _terminalApi.SetColorTableEntry(tableIndex, dwColor);
 }
+
```
Comment (miniksa) [inline src/cascadia/TerminalCore/TerminalDispatch.cpp]: nit: two newlines between functions

File: src/host/outputStream.cpp [hunk 14]
```diff
@@ -761,3 +761,27 @@ BOOL ConhostInternalGetSet::PrivateSetColorTableEntry(const short index, const C
...
+// - value: the new RGB value to use, as a COLORREF, format 0x00BBGGRR. 
+// Return Value:
+// - TRUE if successful (see DoSrvPrivateSetDefaultForegroundColor). FALSE otherwise.
+BOOL ConhostInternalGetSet::PrivateSetDefaultForeground(const COLORREF value) const noexcept
+{
+    return SUCCEEDED(DoSrvPrivateSetDefaultForegroundColor(value));
+}
+
+// Method Description:
+// - Connects the PrivateSetColorTableEntry call directly into our Driver Message servicing
```
Comment (miniksa) [inline src/host/outputStream.cpp]: I think your copy/paste missed updating this comment.

File: src/host/ut_host/ScreenBufferTests.cpp [hunk 15]
```diff
@@ -2502,6 +2506,108 @@ void ScreenBufferTests::SetColorTableThreeDigits()
...
+
+    StateMachine& stateMachine = mainBuffer.GetStateMachine();
+
+    COLORREF originalColor = gci.GetDefaultForegroundColor();
+    COLORREF newColor = gci.GetDefaultForegroundColor();
+    COLORREF testColor = RGB(0x33, 0x66, 0x99);
+    VERIFY_ARE_NOT_EQUAL(originalColor, testColor);
+
+    // Valid hexadecimal notation
+    std::wstring seq = L"\x1b]10;rgb:33/66/99\x1b\\";
```
Comment (miniksa) [inline src/host/ut_host/ScreenBufferTests.cpp]: To me, the exciting part about hex notation is using A-F somewhere. I'm concerned using only 0-9 digits misses the opportunity to guarantee that it is actually parsing hex-wise and not decimal-wise

File: src/host/ut_host/ScreenBufferTests.cpp [hunk 16]
```diff
@@ -2502,6 +2506,108 @@ void ScreenBufferTests::SetColorTableThreeDigits()
...
+    VERIFY_IS_TRUE(WI_IsFlagSet(mainBuffer.OutputMode, ENABLE_VIRTUAL_TERMINAL_PROCESSING));
+
+    StateMachine& stateMachine = mainBuffer.GetStateMachine();
+
+    COLORREF originalColor = gci.GetDefaultForegroundColor();
+    COLORREF newColor = gci.GetDefaultForegroundColor();
+    COLORREF testColor = RGB(0x33, 0x66, 0x99);
+    VERIFY_ARE_NOT_EQUAL(originalColor, testColor);
+
+    // Valid hexadecimal notation
```
Comment (miniksa) [inline src/host/ut_host/ScreenBufferTests.cpp]: Frequently I make these comments be `Log::Comment(L"Valid Hexadecimal Notation")` instead so you can see the sub-tests in the output logs instead of mentally matching up the 4th verify with a particular subtest. Something to consider.

File: src/terminal/adapter/adaptDispatch.cpp [hunk 17]
```diff
@@ -1853,6 +1852,56 @@ bool AdaptDispatch::SetColorTableEntry(const size_t tableIndex,
     return fSuccess;
 }
 
+// Method Description:
+// - Sets the default foreground color to a new value
+// Arguments:
+// - dwColor: The new RGB color value to use, as a COLORREF, format 0x00BBGGRR.
+// Return Value:
+// True if handled successfully. False othewise.
```
Comment (miniksa) [inline src/terminal/adapter/adaptDispatch.cpp]: There's a typo here in `otherwise` and below and probably from wherever this was copied from.

File: src/terminal/adapter/adaptDispatch.hpp [hunk 18]
```diff
@@ -96,8 +96,10 @@ namespace Microsoft::Console::VirtualTerminal
         virtual bool SetCursorStyle(const DispatchTypes::CursorStyle cursorStyle); // DECSCUSR
         virtual bool SetCursorColor(const COLORREF cursorColor);
 
-        virtual bool SetColorTableEntry(const size_t tableIndex,
-                                        const DWORD dwColor); // OscColorTable
+        virtual bool SetColorTableEntry(const size_t tableIndex, const DWORD dwColor); // OscColorTable
+        virtual bool SetDefaultForeground(const DWORD /*dwColor*/); // OSCDefaultForeground
```
Comment (miniksa) [inline src/terminal/adapter/adaptDispatch.hpp]: Why comment the params? None of the rest of the declarations in this file did that.

Comment (miniksa) [reply src/terminal/adapter/adaptDispatch.hpp]: Oh, maybe a copy from the other interface below... still... probably should match file consistency and file another issue to update the overall project style to be more consistent between files.

File: src/terminal/adapter/ut_adapter/adapterTest.cpp [hunk 19]
```diff
@@ -1378,6 +1400,8 @@ class TestGetSet final : public ConGetSet
     short _expectedColorTableIndex = -1;
     COLORREF _expectedColorValue = INVALID_COLOR;
 
+    bool _fPrivateSetDefaultColorResult = false;
```
Comment (miniksa) [inline src/terminal/adapter/ut_adapter/adapterTest.cpp]: I'm not a fan of the overloaded usage of this. I think there's a boolean defined per callback mock with no sharing.

File: src/terminal/parser/OutputStateMachineEngine.cpp [hunk 20]
```diff
@@ -1162,7 +1172,7 @@ bool OutputStateMachineEngine::_VerifyDeviceAttributesParams(_In_reads_(cParams)
 // - Null terminates, then returns, the string that we've collected as part of the OSC string.
 // Arguments:
 // - ppwchTitle - a pointer to point to the Osc String to use as a title.
-// - pcchTitleLength - a pointer place the length of ppwchTitle into.
+// - pcchTitle  - a pointer place the length of ppwchTitle into.
```
Comment (miniksa) [inline src/terminal/parser/OutputStateMachineEngine.cpp]: Yuck. We should file a follow on task to turn all these old-school pointer/size things into `std::wstring_view` and `std::string_view` throughout the parser/adapter...

File: src/terminal/parser/OutputStateMachineEngine.hpp [hunk 21]
```diff
@@ -251,7 +256,7 @@ namespace Microsoft::Console::VirtualTerminal
                                      const size_t cchBuffer,
                                      _Out_ DWORD* const pRgb);
 
-        bool _GetOscSetCursorColor(_In_reads_(cchOscString) const wchar_t* const pwchOscStringBuffer,
+        bool _GetOscSetColor(_In_reads_(cchOscString) const wchar_t* const pwchOscStringBuffer,
                                    const size_t cchOscString,
```
Comment (miniksa) [inline src/terminal/parser/OutputStateMachineEngine.hpp]: nit: it looks like the indentation of the two lines following are now misaligned with the opening parenthesis

File: src/host/outputStream.cpp (hunk 14 above)
Comment (Jaykul) [reply src/host/outputStream.cpp]: Oops 😊

File: src/host/ut_host/ScreenBufferTests.cpp (hunk 15 above)
Comment (Jaykul) [reply src/host/ut_host/ScreenBufferTests.cpp]: Sure 😀 -- My concern is basically just getting the 10 and 11 part right, because the rest of it is just using parsing that already existed, but I'll add another example here.

File: src/host/ut_host/ScreenBufferTests.cpp (hunk 16 above)
Comment (Jaykul) [reply src/host/ut_host/ScreenBufferTests.cpp]: I've literally never written unit tests in C++ and it took me 15 minutes to figure out how to even run these (and I still don't know how to get them to run in Visual Studio), that tip is appreciated.

File: src/terminal/adapter/adaptDispatch.cpp (hunk 17 above)
Comment (Jaykul) [reply src/terminal/adapter/adaptDispatch.cpp]: Yeah, I noticed that same typo elsewhere just now, I'll fix this one too.

Comment (Jaykul) [reply src/terminal/adapter/adaptDispatch.cpp]: LOL It's spelled that way 21 times this file. Fixed.

File: src/terminal/adapter/adaptDispatch.hpp (hunk 18 above)
Comment (Jaykul) [reply src/terminal/adapter/adaptDispatch.hpp]: Oh, yeah ... this is another one of those: "it's been 12 years since I worked in C++" things: if you don't know what you're doing, copy/paste carefully. I will make these match this file.

File: src/terminal/adapter/ut_adapter/adapterTest.cpp (hunk 19 above)
Comment (Jaykul) [reply src/terminal/adapter/ut_adapter/adapterTest.cpp]: I'll fix the duplicated use of _expectedColorValue too.
Are these things actually used anywhere?

File: src/terminal/parser/OutputStateMachineEngine.cpp (hunk 20 above)
Comment (Jaykul) [reply src/terminal/parser/OutputStateMachineEngine.cpp]: I supposed this was easier to build one char at a time ... I assume you'll file a follow up?

**Comment by Jaykul** [comment]
Rebased on current master (to handle merge issues)

File: src/terminal/parser/OutputStateMachineEngine.cpp (hunk 20 above)
Comment (zadjii-msft) [reply src/terminal/parser/OutputStateMachineEngine.cpp]: #975 should take care of this follow-up :)

File: src/terminal/adapter/ut_adapter/adapterTest.cpp (hunk 19 above)
Comment (miniksa) [reply src/terminal/adapter/ut_adapter/adapterTest.cpp]: They might not be. When I stood up the original parser years ago and wrote the mock, every mock was checked somehow. Over time, it probably decayed and the only thing left is that the mock has to be implemented to satisfy compilation of the interface. I'll file a follow-on to try to improve the test coverage overall here.

Comment (miniksa) [reply src/terminal/adapter/ut_adapter/adapterTest.cpp]: Now #987

File: src/terminal/parser/OutputStateMachineEngine.cpp (hunk 9 above)
Comment (fpqc) [reply src/terminal/parser/OutputStateMachineEngine.cpp]: @zadjii-msft I was poking through OutputStateMachineEngine.cpp, and it looks like this code does not actually do anything yet, since the first switch (on line 689 in the main branch as of today) reads:

```
    switch (sOscParam)
    {
    case OscActionCodes::SetIconAndWindowTitle:
    case OscActionCodes::SetWindowIcon:
    case OscActionCodes::SetWindowTitle:
        fSuccess = _GetOscTitle(pwchOscStringBuffer, cchOscString, &pwchTitle, &sCchTitleLength);
        break;
    case OscActionCodes::SetColor:
        fSuccess = _GetOscSetColorTable(pwchOscStringBuffer, cchOscString, &tableIndex, &dwColor);
        break;
    case OscActionCodes::SetForegroundColor:
    case OscActionCodes::SetBackgroundColor:
    case OscActionCodes::SetCursorColor:
        fSuccess = _GetOscSetColor(pwchOscStringBuffer, cchOscString, &dwColor);
        break;
    case OscActionCodes::ResetCursorColor:
        // the console uses 0xffffffff as an "invalid color" value
        dwColor = 0xffffffff;
        fSuccess = true;
        break;
    default:
        // If no functions to call, overall dispatch was a failure.
        fSuccess = false;
        break;
    }
```
Since `fSuccess` is initialized as `false`, this seems like it makes the next switch, which is wrapped in an `if` depending on `fSuccess` skip the OSC 10 and 11 cases:
```
    if (fSuccess)
    {
        switch (sOscParam)
        {
        case OscActionCodes::SetIconAndWindowTitle:
        case OscActionCodes::SetWindowIcon:
        case OscActionCodes::SetWindowTitle:
            fSuccess = _dispatch->SetWindowTitle({ pwchTitle, sCchTitleLength });
            TermTelemetry::Instance().Log(TermTelemetry::Codes::OSCWT);
            break;
        case OscActionCodes::SetColor:
            fSuccess = _dispatch->SetColorTableEntry(tableIndex, dwColor);
            TermTelemetry::Instance().Log(TermTelemetry::Codes::OSCCT);
            break;
        case OscActionCodes::SetForegroundColor:
            fSuccess = _dispatch->SetDefaultForeground(dwColor);
            TermTelemetry::Instance().Log(TermTelemetry::Codes::OSCFG);
            break;
        case OscActionCodes::SetBackgroundColor:
            fSuccess = _dispatch->SetDefaultBackground(dwColor);
            TermTelemetry::Instance().Log(TermTelemetry::Codes::OSCBG);
            break;
        case OscActionCodes::SetCursorColor:
            fSuccess = _dispatch->SetCursorColor(dwColor);
            TermTelemetry::Instance().Log(TermTelemetry::Codes::OSCSCC);
            break;
        case OscActionCodes::ResetCursorColor:
            fSuccess = _dispatch->SetCursorColor(dwColor);
            TermTelemetry::Instance().Log(TermTelemetry::Codes::OSCRCC);
            break;
        default:
            // If no functions to call, overall dispatch was a failure.
            fSuccess = false;
            break;
        }
    }
```

Doesn't this mean that the code in this PR right now isn't actually doing anything, since it's always skipped over?  Should `fSuccess` instead be initialized as `true`?

Comment (DHowett-MSFT) [reply src/terminal/parser/OutputStateMachineEngine.cpp]: The case additions [here](https://github.com/microsoft/terminal/pull/891/files#diff-6ddb7535cb6d02274d60e614e3780e4dR700) should make sure we _do_ dispatch to `_GetOscSetColor`, if I'm reading it properly.

Comment (fpqc) [reply src/terminal/parser/OutputStateMachineEngine.cpp]: @DHowett Sorry if I'm being dense, but when you initialize the variable `fsuccess` for the method, you initialize it as false, rather than true.  Therefore, those case additions don't help. See line 683 in the current master branch.

Comment (DHowett-MSFT) [reply src/terminal/parser/OutputStateMachineEngine.cpp]: https://github.com/microsoft/terminal/blob/94bcbb92045f9daf16168cd82b1abdf731065bc8/src/terminal/parser/OutputStateMachineEngine.cpp#L683-L703

The switch on 689 has a case at 699,700 that covers these Osc codes, and that changes `fSuccess`.

Sure is! Select a line or block of lines (with shift), click the `...` and choose _copy permalink_. Just pasting that URL will make github insert an inline code block.

image

image

Comment (fpqc) [reply src/terminal/parser/OutputStateMachineEngine.cpp]: Ok, sorry, one last question, since I'm hoping to try to implement another OSC code (52) in the future:  Why does that switch case automagically set `fSuccess` to true without having to add
```
fsuccess = true;
break;
```

Comment (DHowett-MSFT) [reply src/terminal/parser/OutputStateMachineEngine.cpp]: `fSuccess = _GetOscSetColor(...)`, and `_GetOscSetColor` returns whether _it_ succeeded.

Comment (DHowett-MSFT) [reply src/terminal/parser/OutputStateMachineEngine.cpp]: In the future, you’ll wonder why the entire exchange ultimately does _fail_. Getting a false out of the last step in the dispatch loop means “and also send it over the pseudoconsole, because I didn’t handle it to a state that is complete enough for a connected terminal.” That’s why you will see some “if in VtIo mode, return false” exchanges.

Comment (fpqc) [reply src/terminal/parser/OutputStateMachineEngine.cpp]: @DHowett Ohhhh, I think I see, those cases get combined in C++ if not followed by a break.  I assumed there was always an implicit break between cases, but it's actually working like a goto and will execute all operations until the next break. Good to know! Thanks!

**Comment by werner-1** [comment]
Sorry if this is the wrong place to ask, but how do I use this?
I tried printf %b '\033]11;#200000\007' in wsl

**Comment by DHowett** [comment]
The color format that the xterm color OSCs take is certainly strange :smile: it's `rgb:RR/GG/BB` in hex.

You'll want to try `printf '\033]11;rgb:20/00/00\a'`
